src/sqlfluff/core/parser/grammar/delimited.py

```diff
@@ -182,3 +182,53 @@ def match(
             return MatchResult.empty_at(idx)
 
         return working_match
+
+
+class OptionallyDelimited(AnyNumberOf):
+    """Match a number of elements optionally separated by a delimiter.
+
+    Note that if there are multiple elements passed in that they will be treated
+    as different options of what can be delimited, rather than a sequence.
+    """
+
+    equality_kwargs: tuple[str, ...] = (
+        "_elements",
+        "optional",
+        "allow_gaps",
+        "delimiter",
+        "allow_trailing",
+        "terminator",
+        "min_delimiters",
+    )
+
+    def __init__(
+        self,
+        *args: Union[Matchable, str],
+        delimiter: Union[Matchable, str] = Ref("CommaSegment"),
+        allow_trailing: bool = False,
+        terminators: Sequence[Union[Matchable, str]] = (),
+        reset_terminators: bool = False,
+        min_delimiters: int = 0,
+        bracket_pairs_set: str = "bracket_pairs",
+        allow_gaps: bool = True,
+        optional: bool = False,
+    ) -> None:
+        """Initialize the class object with the provided arguments."""
+        super().__init__(
+            *args,
+            Delimited(
+                *args,
+                delimiter=delimiter,
+                allow_trailing=allow_trailing,
+                terminators=terminators,
+                reset_terminators=reset_terminators,
+                min_delimiters=min_delimiters,
+                bracket_pairs_set=bracket_pairs_set,
+                allow_gaps=allow_gaps,
+                optional=optional,
+            ),
+            terminators=terminators,
+            reset_terminators=reset_terminators,
+            allow_gaps=allow_gaps,
+            optional=optional,
```

Comment:
Doing it this way means that we can now parse:
```sql
ALTER USER "John" SET DEFAULT_ROLE = JOHNNY DEFAULT_WAREHOUSE = FEU COMMENT = "Foo";
ALTER USER "John" SET DEFAULT_ROLE = JOHNNY, DEFAULT_WAREHOUSE = FEU, COMMENT = "Foo";
```

But if there's a mix of commas and space delimiters, then it will fail 🤔 . e.g.

```sql
ALTER USER "John" SET DEFAULT_ROLE = JOHNNY DEFAULT_WAREHOUSE = FEU, COMMENT = "Foo";
```

I realise it's an edge case in an edge case, but I wonder if we need a different approach. In the `Delimited` grammar, the section of code which looks for delimiters or content is fairly compact, and it probably wouldn't be crazy to extent that to search optionally for both if configured that way.